src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -375,4 +375,706 @@ namespace OpenMS
+      {
+        std::cout << d << " ";
+      }
+      std::cout << std::endl;
+      std::cout << "result=" << result << std::endl;
+    }
+    return result;
+  }
+
+  double PeakIntegrator::Loss_function(
+    const std::vector<double>& xs,
+    const std::vector<double>& ys,
+    const double h,
+    const double mu,
+    const double sigma,
+    const double tau
+  ) const
+  {
+    const double u = mu;
+    const double s = sigma;
+    const double t = tau;
+    std::vector<double> diffs(xs.size());
+    for (Size i = 0; i < xs.size(); ++i)
+    {
+      const double x = xs[i];
+      const double y = ys[i];
+      const double z = compute_z(x, mu, sigma, tau);
+      if (z < 0)
+      {
+        diffs[i] = (std::pow(((((h*s)/t) * std::sqrt(PI/2.0) * std::exp((1.0/2.0)*(std::pow(s/t,2.0))-(x-u)/t) * std::erfc((1/std::sqrt(2.0)) * (s/t - (x-u)/s)))-y),2.0)) / static_cast<double>(xs.size());
+      }
+      else if (z <= 6.71e7)
+      {
+        diffs[i] = (std::pow((h * std::exp(-1.0/2.0 * std::pow(((x - u)/s),2.0)) * (s/t) * std::sqrt(PI/2.0) * std::exp(std::pow((1.0/std::sqrt(2.0) * (s/t - (x - u)/s)),2.0)) * std::erfc(1.0/std::sqrt(2.0) * (s/t - (x - u)/s)) - y),2.0)) / static_cast<double>(xs.size());
+      }
+      else
+      {
+        diffs[i] = (std::pow((((h * std::exp(-(1.0/2.0) * (std::pow(((x-u) / s),2.0)))) / (  1.0 - (((x-u) * t) / (std::pow(s,2.0)))))-y),2.0)) / static_cast<double>(xs.size());
+      }
+    }
+    const double result = std::accumulate(diffs.begin(), diffs.end(), 0.0);
+    if (print_debug_ == 2)
+    {
+      std::cout << std::endl << "Loss_function() diffs:" << std::endl;
+      for (double d : diffs)
+      {
+        std::cout << d << " ";
+      }
+      std::cout << std::endl;
+      std::cout << "result=" << result << std::endl;
+    }
+    return result;
+  }
+
+  void PeakIntegrator::emg_vector(
+    const std::vector<double>& xs,
+    const double h,
+    const double mu,
+    const double sigma,
+    const double tau,
+    std::vector<double>& out_xs,
+    std::vector<double>& out_ys,
+    const bool compute_additional_points
+  ) const
+  {
+    out_xs = xs; // Copy all positions to output
+    for (const double x : out_xs) // For each x, estimate y
+    {
+      out_ys.push_back(emg_point(x, h, mu, sigma, tau));
+    }
+
+    if (!compute_additional_points) return;
+
+    // Compute the sampling step for the additional points
+    double avg_sampling { 0.0 };
+    for (Size i = 1; i < xs.size(); ++i)
+    {
+      avg_sampling += xs[i] - xs[i - 1];
+    }
+    avg_sampling /= xs.size() - 1;
+
+    // Stop adding points if the estimated y <= `est_y_threshold`
+    const double est_y_threshold { 1e-3 };
+
+    // Decide on which side the eventual additional points should be added
+    // The loop stops once the last added point's intensity is:
+    // - lower than the intensity on the opposite boundary
+    // - lower than `est_y_threshold`
+    if (out_ys.front() > out_ys.back())
```

Comment:
Are there any checks for a case where the interpolated points over extend the baseline of the peak?  I can imagine a situation where missing points on the tailing side of the peak could potentially be extended out to infinity.  Are there any checks on the maximum baseline?

Reply:
I noticed cases where many points would be added (eg. the peak would have 4 times the number of original points).
It would help me to see a visualization of the problem to understand it and, if needed, solve in some way.
No infinite case was experienced with the sample you had provided me, though :smiley:

Reply:
OK lets discuss this check more during our meeting today.

Reply:
TODO: use 3 times the width of the "valid" peak's half